src/evo/deterministicmns.cpp

```diff
@@ -708,9 +708,50 @@ bool CDeterministicMNManager::BuildNewListFromBlock(const CBlock& block, const C
                 LogPrintf("CDeterministicMNManager::%s -- MN %s added at height %d: %s\n",
                     __func__, tx.GetHash().ToString(), nHeight, pl.ToString());
             }
+
+        } else if (tx.nType == CTransaction::TxType::PROUPSERV) {
+            ProUpServPL pl;
+            if (!GetTxPayload(tx, pl)) {
+                return _state.DoS(100, false, REJECT_INVALID, "bad-protx-payload");
+            }
+
+            if (newList.HasUniqueProperty(pl.addr) && newList.GetUniquePropertyMN(pl.addr)->proTxHash != pl.proTxHash) {
+                return _state.DoS(100, false, REJECT_DUPLICATE, "bad-protx-dup-addr");
+            }
+
+            CDeterministicMNCPtr dmn = newList.GetMN(pl.proTxHash);
+            if (!dmn) {
+                return _state.DoS(100, false, REJECT_INVALID, "bad-protx-hash");
+            }
+            if (dmn->nOperatorReward == 0 && !pl.scriptOperatorPayout.empty()) {
+                // operator payout address can not be set if the operator reward is 0
+                return _state.DoS(100, false, REJECT_INVALID, "bad-protx-operator-payee");
+            }
+            auto newState = std::make_shared<CDeterministicMNState>(*dmn->pdmnState);
+            newState->addr = pl.addr;
+            newState->scriptOperatorPayout = pl.scriptOperatorPayout;
+
+            if (newState->nPoSeBanHeight != -1) {
+                // only revive when all keys are set
+                if (!newState->keyIDOperator.IsNull() && !newState->keyIDVoting.IsNull() && !newState->keyIDOwner.IsNull()) {
+                    newState->nPoSePenalty = 0;
+                    newState->nPoSeBanHeight = -1;
+                    newState->nPoSeRevivedHeight = nHeight;
```

Comment:
As far as I'm seeing, the `PROUPSERV` does not update keys, only the service address and the payout script. So, how `keyIDOperator`, `keyIDOwner` or `keyIDVoting` can be null and be updated by this payload then?

Reply:
Yes, this transaction does not update any key.
This code is only checking that all keys are *already* set (in the previous list) before "reviving" a banned masternode (masternode revival is outlined in the [DIP3](https://github.com/dashpay/dips/blob/master/dip-0003.md) document, as well as in #2267 description).

For example, the masternode could have had its operator key revoked by a (prior) ProUpRev transaction (#2363). In this case, the masternode is not revived by a ProUpServ, until a ProUpReg transaction sets the new operator key.

Checking also owner and voting key is indeed redundant, but it doesn't hurt.

Reply:
k good.